# Goals

The pipeline we ran through last time took you from raw next gen data to a sorted bam file for each sample. Before we genotype these data, we need to do two more steps:
* Deduping reads and
* Realigning indels

After that, we will:
* genotype our data and
* Filter genotypes

# Dedupling a bam file

Many approaches to library construction have a PCR step, which means that some data you generate may have the same start and stop points for both pairs of a read.  This causes an overestimation of confidence and can lead to miscalled genotypes if one allele amplified more than the other.  Typically we therefore remove duplicates ("dedup") before genotyping.  We didn't do this last time (for the sake of time) but we are gonna do it now.

I've prepared an example `.bam` file that I'd like us to work with.  Please make a symbolic link to this file:
```
ln -s /2/scratch/Bio722_BJE/PM561_chr9_sorted.bam
```


